src/Mod/Part/App/TopoShapeExpansion.cpp

```diff
@@ -509,6 +514,743 @@ void TopoShape::mapSubElement(const std::vector<TopoShape>& shapes, const char*
     }
 }
 
+struct ShapeInfo
+{
+    const TopoDS_Shape& shape;
+    TopoShapeCache::Ancestry& cache;
+    TopAbs_ShapeEnum type;
+    const char* shapetype;
+
+    ShapeInfo(const TopoDS_Shape& shape, TopAbs_ShapeEnum type, TopoShapeCache::Ancestry& cache)
+        : shape(shape)
+        , cache(cache)
+        , type(type)
+        , shapetype(TopoShape::shapeName(type).c_str())
+    {}
+
+    int count() const
+    {
+        return cache.count();
+    }
+
+    TopoDS_Shape find(int index)
+    {
+        return cache.find(shape, index);
+    }
+
+    int find(const TopoDS_Shape& subshape)
+    {
+        return cache.find(shape, subshape);
+    }
+};
+
+////////////////////////////////////////
+// makESHAPE -> makeShapeWithElementMap
+///////////////////////////////////////
+
+struct NameKey
+{
+    Data::MappedName name;
+    long tag = 0;
+    int shapetype = 0;
+
+    NameKey()
+    = default;
+    explicit NameKey(Data::MappedName  n)
+        : name(std::move(n))
+    {}
+    NameKey(int type, Data::MappedName  n)
+        : name(std::move(n))
+    {
+        // Order the shape type from vertex < edge < face < other.  We'll rely
+        // on this for sorting when we name the geometry element.
+        switch (type) {
+            case TopAbs_VERTEX:
+                shapetype = 0;
+                break;
+            case TopAbs_EDGE:
+                shapetype = 1;
+                break;
+            case TopAbs_FACE:
+                shapetype = 2;
+                break;
+            default:
+                shapetype = 3;
+        }
+    }
+    bool operator<(const NameKey& other) const
+    {
+        if (shapetype < other.shapetype) {
+            return true;
+        }
+        if (shapetype > other.shapetype) {
+            return false;
+        }
+        if (tag < other.tag) {
+            return true;
+        }
+        if (tag > other.tag) {
+            return false;
+        }
+        return name < other.name;
+    }
+};
+
+struct NameInfo
+{
+    int index {};
+    Data::ElementIDRefs sids;
+    const char* shapetype {};
+};
+
+
+const std::string& modPostfix()
+{
+    static std::string postfix(TopoShape::elementMapPrefix() + ":M");
+    return postfix;
+}
+
+const std::string& modgenPostfix()
+{
+    static std::string postfix(modPostfix() + "G");
+    return postfix;
+}
+
+const std::string& genPostfix()
+{
+    static std::string postfix(TopoShape::elementMapPrefix() + ":G");
+    return postfix;
+}
+
+const std::string& upperPostfix()
+{
+    static std::string postfix(TopoShape::elementMapPrefix() + ":U");
+    return postfix;
+}
+
+const std::string& lowerPostfix()
+{
+    static std::string postfix(TopoShape::elementMapPrefix() + ":L");
+    return postfix;
+}
+
```

Comment:
is it the same as this https://github.com/FreeCAD/FreeCAD/blob/main/src/App/ElementNamingUtils.h?

Reply:
Yeah, I thought I had gone back and referenced the constants, but apparently not.

Reply:
OK, that's done now.